From d1bc23f339c10b75ec0149baf54e4f37e4f393a1 Mon Sep 17 00:00:00 2001
From: Samreen Nilofer <samreen.nilofer@intel.com>
Date: Wed, 23 Apr 2014 14:50:41 +0530
Subject: Audio: sst: Check payload size for large messages

BZ: 190729

LPE FW, sends a large message with zero payload length.
Check the payload size if the message is large.

Change-Id: I12861a6fb6c50d40094e46bf3be2ef764644a740
Signed-off-by: Samreen Nilofer <samreen.nilofer@intel.com>
---
 sound/soc/intel/sst/sst_ipc.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/sound/soc/intel/sst/sst_ipc.c b/sound/soc/intel/sst/sst_ipc.c
index b005c53..517fd05 100644
--- a/sound/soc/intel/sst/sst_ipc.c
+++ b/sound/soc/intel/sst/sst_ipc.c
@@ -691,6 +691,19 @@ void sst_process_reply_mrfld(struct ipc_post *msg)
 	/* if it is a large message, the payload contains the size to
 	 * copy from mailbox */
 	if (msg_high.part.large) {
+
+		if (!msg_low) {
+			pr_err("payload size is 0 for large message\n");
+			pr_err("IPC header %#x has %#x payload\n",
+					msg_high.full, msg_low);
+
+			sst_wake_up_block(sst_drv_ctx, msg_high.part.result,
+					msg_high.part.drv_id,
+					msg_high.part.msg_id, NULL, 0);
+			WARN_ON(1);
+			return;
+		}
+
 		data = kzalloc(msg_low, GFP_KERNEL);
 		if (!data)
 			goto end;
-- 
2.0.0

